_posts/2018-02-12-Reusable-features-in-React-Redux.md

```diff
@@ -0,0 +1,188 @@
+     TodoListItem.js
+   action.js
+   reducers.js
+   selectors.js
+   utils.js
+   middleware.js
+```   
+
+#### components/
+
+The directory where your components live.
+
+#### containers/
+
+The directory where your containers live. If you don't know what is the difference between the components 
+and containers you should read [this article](https://medium.com/@dan_abramov/smart-and-dumb-components-7ca2f9a7c7d0)
+from Dan Abramov why this separation exists and why it is necessary.
+
+#### actions.js
+
+The file where your [action types](https://redux.js.org/docs/basics/Actions.html#actions) and
+[action creators](https://redux.js.org/docs/basics/Actions.html#action-creators) are defined.
+
+
+#### reducers.js
+
+The file where your [reducers](https://redux.js.org/docs/basics/Reducers.html) are defined. 
+This is also a place where you compose the reducers of your possible sub-features using [combineReducers](https://redux.js.org/docs/api/combineReducers.html).
+
+
+#### selectors.js
+
+The file where you define your selectors. Selectors are pure functions that compose and know how
+to retrieve data from your state. If you're now familiar with the concept I highly recommend looking 
+at [reselect](https://github.com/reactjs/reselect) library. I strongly recommend dividing the selectors into two categories. 
+Input-selectors that only retrieve data from the state and combined-selectors that can combine
+multiple input-selectors and add some business logic or do some transformations.
+
+
+#### utils.js
+
+Everything related to your feature that doesn't belong anywhere else. Usually it contains support utilities and helpers.
+
+#### middleware.js
+
+This is where our feature handles it's side effects, e.g. catching some Redux actions and dispatching new ones,
+synchronize the API calls, etc...
+
+You can check the [reference repository](https://github.com/char0n/react-reusable-features/tree/master/src/app/standard-feature) to see how this looks in a real project. 
+It depends on your use-case what the standard feature directory contains. 
+There are situations that you may need reducers but don't need actions and so on...
+But regardless, the full representation of the standard feature should look like that. 
+If the number of the components in the standard feature becomes too big, you break the standard
+feature into more sub-features which also adheres to the same rules as it's parent feature.     
+
+
+```
+src/
+ todo-feature/
+   components/ 
+     TodoList.js
+   containers/
+     TodoList.js
+   todo-item/
+     components/
+       TodoListItem.js
+     containers/
+       TodoListItem.js
+     actions.js
+     selectors.js 
+   action.js
+   reducers.js
+   selectors.js
+   utils.js
+   middleware.js
+```
+
+
+It is always good to break features into smaller sub-features with encapsulated logic to reduce the complexity.
+Try to think of it as applying Single Responsibility Principle. One feature does only one thing and
```

Comment:
*The* Single Responsibility Principle. I'd perhaps contain a hyperlink so the user can read more if this is an unfamiliar term.

```diff
- Try to think of it as applying Single Responsibility Principle.
+ Try to think of it as applying The Single Responsibility Principle.
```